File: crates/flowrt-codegen/tests/golden/cross_process_feedback_cpp/expected/cpp/tests/message_abi.cpp
// FlowRT 管理产物。不要手工修改。

#include <array>
#include <cassert>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <type_traits>

#include "flowrt_app/messages.hpp"

namespace {

template <typename T>
std::array<std::uint8_t, sizeof(T)> bytes_of(const T& value) {
    std::array<std::uint8_t, sizeof(T)> bytes{};
    std::memcpy(bytes.data(), &value, bytes.size());
    return bytes;
}

template <typename T>
void assert_default_bytes_zero() {
    T value{};
    std::array<std::uint8_t, sizeof(T)> expected{};
    assert(bytes_of(value) == expected);
}

template <typename T>
void assert_byte_roundtrip(const T& value) {
    const auto bytes = bytes_of(value);
    T roundtrip{};
    std::memset(&roundtrip, 0, sizeof(roundtrip));
    std::memcpy(&roundtrip, bytes.data(), bytes.size());
    assert(std::memcmp(&roundtrip, &value, sizeof(T)) == 0);
}

template <typename T, std::size_t N>
void assert_sample_bytes(const T& value, const std::array<std::uint8_t, N>& expected) {
    static_assert(sizeof(T) == N);
    assert(bytes_of(value) == expected);
}

template <std::size_t N>
void write_fixture(std::string_view name, const std::array<std::uint8_t, N>& bytes) {
#ifdef FLOWRT_ABI_FIXTURE_DIR
    std::filesystem::create_directories(FLOWRT_ABI_FIXTURE_DIR);
    auto path = std::filesystem::path(FLOWRT_ABI_FIXTURE_DIR) / std::string(name);
    std::ofstream output(path, std::ios::binary);
    if (!output) {
        throw std::runtime_error("failed to open ABI fixture output");
    }
    output.write(reinterpret_cast<const char*>(bytes.data()), static_cast<std::streamsize>(bytes.size()));
    if (!output) {
        throw std::runtime_error("failed to write ABI fixture output");
    }
#else
    (void)name;
    (void)bytes;
#endif
}

constexpr std::array<std::uint8_t, 8> EXPECTED_CMD_BYTES{{0, 0, 0, 0, 0, 0, 2, 64}};
constexpr std::array<std::uint8_t, 8> EXPECTED_STATE_BYTES{{0, 0, 0, 0, 0, 0, 2, 64}};

flowrt_app::Cmd sample_cmd() {
    flowrt_app::Cmd value{};
    std::memset(&value, 0, sizeof(value));
    value.u = 2.25;
    return value;
}

flowrt_app::State sample_state() {
    flowrt_app::State value{};
    std::memset(&value, 0, sizeof(value));
    value.x = 2.25;
    return value;
}

void test_cmd_message_abi() {
    static_assert(std::is_standard_layout_v<flowrt_app::Cmd>);
    static_assert(std::is_trivially_copyable_v<flowrt_app::Cmd>);
    static_assert(sizeof(flowrt_app::Cmd) == 8);
    static_assert(alignof(flowrt_app::Cmd) == 8);
    assert_default_bytes_zero<flowrt_app::Cmd>();
    static_assert(offsetof(flowrt_app::Cmd, u) == 0);
    assert_byte_roundtrip(sample_cmd());
    assert_sample_bytes(sample_cmd(), EXPECTED_CMD_BYTES);
    write_fixture("cmd.bin", bytes_of(sample_cmd()));
}

void test_cmd_wire_codec_omits_native_padding() {
    const auto value = sample_cmd();
    std::array<std::uint8_t, flowrt_app::Cmd::wire_size()> wire{};
    value.encode_wire(wire);
    const std::array<std::uint8_t, flowrt_app::Cmd::wire_size()> expected_wire{0, 0, 0, 0, 0, 0, 2, 64};
    assert(wire == expected_wire);
    const auto decoded = flowrt_app::Cmd::decode_wire(wire);
    assert(bytes_of(decoded) == bytes_of(value));
}

void test_state_message_abi() {
    static_assert(std::is_standard_layout_v<flowrt_app::State>);
    static_assert(std::is_trivially_copyable_v<flowrt_app::State>);
    static_assert(sizeof(flowrt_app::State) == 8);
    static_assert(alignof(flowrt_app::State) == 8);
    assert_default_bytes_zero<flowrt_app::State>();
    static_assert(offsetof(flowrt_app::State, x) == 0);
    assert_byte_roundtrip(sample_state());
    assert_sample_bytes(sample_state(), EXPECTED_STATE_BYTES);
    write_fixture("state.bin", bytes_of(sample_state()));
}

void test_state_wire_codec_omits_native_padding() {
    const auto value = sample_state();
    std::array<std::uint8_t, flowrt_app::State::wire_size()> wire{};
    value.encode_wire(wire);
    const std::array<std::uint8_t, flowrt_app::State::wire_size()> expected_wire{0, 0, 0, 0, 0, 0, 2, 64};
    assert(wire == expected_wire);
    const auto decoded = flowrt_app::State::decode_wire(wire);
    assert(bytes_of(decoded) == bytes_of(value));
}

}  // namespace

int main() {
    test_cmd_message_abi();
    test_cmd_wire_codec_omits_native_padding();
    test_state_message_abi();
    test_state_wire_codec_omits_native_padding();
    return 0;
}

File: crates/flowrt-codegen/tests/golden/cross_process_feedback_rust/expected/rust/tests/message_abi.rs
// FlowRT 管理产物。不要手工修改。

use flowrt::WireCodec;

fn bytes_of<T>(value: &T) -> Vec<u8> {
    let mut bytes = vec![0u8; std::mem::size_of::<T>()];
    // Safety：生成测试只传入 FlowRT ABI v0.1 plain-data 消息，且 padding 已初始化。
    unsafe {
        std::ptr::copy_nonoverlapping(
            (value as *const T).cast::<u8>(),
            bytes.as_mut_ptr(),
            bytes.len(),
        );
    }
    bytes
}

fn assert_default_bytes_zero<T: Copy + Default>() {
    let value = T::default();
    assert_eq!(bytes_of(&value), vec![0u8; std::mem::size_of::<T>()]);
}

fn assert_byte_roundtrip<T: Copy + Default>(value: T) {
    let bytes = bytes_of(&value);
    let mut roundtrip = T::default();
    // Safety：`roundtrip` 是有效 plain-data 存储，`bytes` 长度等于 `size_of::<T>()`。
    unsafe {
        std::ptr::copy_nonoverlapping(
            bytes.as_ptr(),
            (&mut roundtrip as *mut T).cast::<u8>(),
            bytes.len(),
        );
    }
    assert_eq!(bytes_of(&roundtrip), bytes);
}

fn assert_sample_bytes<T: Copy>(value: T, expected: &[u8]) {
    assert_eq!(bytes_of(&value), expected);
}

const EXPECTED_CMD_BYTES: &[u8] = &[0, 0, 0, 0, 0, 0, 2, 64];
const EXPECTED_STATE_BYTES: &[u8] = &[0, 0, 0, 0, 0, 0, 2, 64];

fn sample_cmd() -> flowrt_app::messages::Cmd {
    let mut value = flowrt_app::messages::Cmd::default();
    value.u = 2.25f64;
    value
}

fn sample_state() -> flowrt_app::messages::State {
    let mut value = flowrt_app::messages::State::default();
    value.x = 2.25f64;
    value
}

#[test]
fn cmd_message_abi() {
    assert_eq!(std::mem::size_of::<flowrt_app::messages::Cmd>(), 8);
    assert_eq!(std::mem::align_of::<flowrt_app::messages::Cmd>(), 8);
    assert_default_bytes_zero::<flowrt_app::messages::Cmd>();
    assert_eq!(std::mem::offset_of!(flowrt_app::messages::Cmd, u), 0);
    assert_byte_roundtrip(sample_cmd());
    assert_sample_bytes(sample_cmd(), EXPECTED_CMD_BYTES);
}

#[test]
fn cmd_wire_codec_omits_native_padding() {
    let value = sample_cmd();
    let wire = value.to_wire_vec().unwrap();
    assert_eq!(wire, vec![0, 0, 0, 0, 0, 0, 2, 64]);
    assert_eq!(flowrt_app::messages::Cmd::decode_wire(&wire).unwrap(), value);
}

#[test]
fn state_message_abi() {
    assert_eq!(std::mem::size_of::<flowrt_app::messages::State>(), 8);
    assert_eq!(std::mem::align_of::<flowrt_app::messages::State>(), 8);
    assert_default_bytes_zero::<flowrt_app::messages::State>();
    assert_eq!(std::mem::offset_of!(flowrt_app::messages::State, x), 0);
    assert_byte_roundtrip(sample_state());
    assert_sample_bytes(sample_state(), EXPECTED_STATE_BYTES);
}

#[test]
fn state_wire_codec_omits_native_padding() {
    let value = sample_state();
    let wire = value.to_wire_vec().unwrap();
    assert_eq!(wire, vec![0, 0, 0, 0, 0, 0, 2, 64]);
    assert_eq!(flowrt_app::messages::State::decode_wire(&wire).unwrap(), value);
}
